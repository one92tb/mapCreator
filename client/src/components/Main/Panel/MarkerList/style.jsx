import styled from "styled-components";

const List = styled.div`
  list-style: none;
  padding: 0;
  height: 100%;
`;

const Marker = styled.li`
  margin-bottom: 5px;
  border: 1px solid #4ddbff;
  width: 100%;
  height: 40px;
  border-radius: 3px;
  margin-bottom: 5px !important;
  padding: 0 !important;
  display: flex;

  &:hover {
    background: #4ddbff;
    cursor: pointer;
  }

  ${({ isSelected, isFiltered }) => {
    if (isSelected) {
      return `background: #00b8e6`;
    } else if (isFiltered) {
      return `background: #999; opacity: 0.7`;
    }
  }};
`;

const MarkerIcon = styled.div`
  margin: 0 10px;
  float: left;
  display: flex;
  justify-content: center;
  align-items: center;
`;

const MarkerName = styled.span`
  width: 75%;
  float: left;
  display: flex;
  align-items: center;
`;

export { List, Marker, MarkerIcon, MarkerName };
