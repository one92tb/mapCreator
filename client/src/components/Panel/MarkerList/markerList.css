.markerBox {
  margin-bottom: 5px;
  border: 1px solid rgba(0, 0, 0, 0.125);
  width: 100%;
  height: 50px;
  border-radius: 3px;
  margin-bottom: 5px!important;
  padding: 0!important;
  display: flex;
}

.markerBox__name {
  width: 75%;
  float: left;
  justify-content: center;
  display: flex;
  border-right: 1px solid rgba(0, 0, 0, 0.03);
  align-items: center;

}

.markerBox__icon {
  width: 25%;
  float: left;
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: rgba(0, 0, 0, 0.125);

}

.selectedMarker{
  background: red;
}
.markerList{
  list-style: none;
  padding: 0;
}
