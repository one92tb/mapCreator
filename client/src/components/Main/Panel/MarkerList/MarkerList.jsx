import React, { Component } from "react";
import { fetchMarkers } from "../../../../actions/marker/fetchMarkers";
import { disableMarkers } from "../../../../actions/marker/disableMarkers";
import { getSelectedMarker } from "../../../../actions/marker/getSelectedMarker";
import { connect } from "react-redux";
import { List, Marker, MarkerIcon, MarkerName, MarkerImg } from "./style";
import PropTypes from "prop-types";
import baseUrl from "../../../../baseUrl";

List.displayName = "div";
Marker.displayName = "li";
MarkerIcon.displayName = "div";
MarkerName.displayName = "span";
MarkerImg.displayName = "img";

export class MarkerList extends Component {
  constructor(props) {
    super(props);
    this.state = {
      selectedId: "",
      filteredMarkers: []
    };
  }

  componentDidMount() {
    const { fetchMarkers } = this.props;
    fetchMarkers();
  }

  componentDidUpdate() {
    const { selectedId } = this.state;
    const { selectedMarker } = this.props;
    // "", { ...} , !== ""

    if (selectedId !== selectedMarker && selectedMarker === "") {
      this.setState({
        selectedId: ""
      });
    }
  }

  onSelect = (marker, id) => {
    const { getSelectedMarker, isNavSelect, disableMarkers } = this.props;
    const { selectedId, filteredMarkers } = this.state;

    if (marker.id !== selectedId && isNavSelect) {
      this.setState({ selectedId: id });
      getSelectedMarker({
        ...marker,
        url: `${baseUrl}/images/${marker.icon}`
      });
    } else if (marker.id === selectedId && isNavSelect) {
      this.setState({ selectedId: "" });
      getSelectedMarker({
        id: undefined,
        name: "",
        url: "IMG-default.png"
      });
    } else {
      //remove Marker from filteredMarkers if exist
      if (filteredMarkers.find(el => el.id === marker.id)) {
        this.setState(
          {
            filteredMarkers: filteredMarkers.filter(el => el.id !== marker.id)
          },
          () => {
            disableMarkers(this.state.filteredMarkers);
          }
        );
        //add Marker to filteredMarkers if not exist
      } else {
        this.setState(
          {
            filteredMarkers: [...filteredMarkers, marker]
          },
          () => {
            disableMarkers(this.state.filteredMarkers);
          }
        );
      }
    }
  };

  render() {
    const { isNavSelect, markers } = this.props;
    const { selectedId, filteredMarkers } = this.state;
    return (
      <List>
        {markers.map((marker, id) => (
          <Marker
            key={marker.id}
            isSelected={selectedId === marker.id && isNavSelect}
            isFiltered={
              filteredMarkers.find(el => el.id === marker.id) && !isNavSelect
            }
            onClick={() => this.onSelect(marker, marker.id)}
          >
            <MarkerIcon>
              <MarkerImg
                src={`${baseUrl}/images/${marker.icon}`}
                alt={marker.icon}
              />
            </MarkerIcon>
            <MarkerName>{marker.name}</MarkerName>
          </Marker>
        ))}
      </List>
    );
  }
}

const mapStateToProps = state => ({
  selectedMarker: state.marker.selectedMarker,
  markers: state.marker.markers
});

const mapDispatchToProps = {
  fetchMarkers,
  getSelectedMarker,
  disableMarkers
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(MarkerList);

MarkerList.propTypes = {
  fetchMarkers: PropTypes.func.isRequired,
  getSelectedMarker: PropTypes.func.isRequired,
  disableMarkers: PropTypes.func.isRequired,
  isNavSelect: PropTypes.bool.isRequired,
  markers: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired,
      icon: PropTypes.string.isRequired,
      userId: PropTypes.number.isRequired
    })
  )
};

MarkerList.defaultProps = {};
