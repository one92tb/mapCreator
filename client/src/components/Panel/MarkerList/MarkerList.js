import React, {Component} from 'react';
import './markerList.css';
import {ListGroup, ListGroupItem} from 'reactstrap';
import {fetchMarkerRecord} from '../../../actions/fetchMarkerRecord';
import {connect} from 'react-redux';

class MarkerList extends Component {

  constructor(props) {
    super(props);
    this.state = {
      selectedId: ''
    };

    this.onSelect = this.onSelect.bind(this);
  }

  componentDidMount(){
    this.props.fetchMarkerRecord();
  }

  onSelect(id) {
    if (id === this.state.selectedId) {
      id = '';
      this.setState({selectedId: id})
    } else {
      this.setState({selectedId: id})
    }
  }

  render() {
    console.log(this.props);
    const initData = [
      {
        name: 'Home',
        icon: 'icon-home'
      }, {
        name: 'Football Pitch',
        icon: 'icon-soccer-ball'
      }
    ]

    const {selectedId} = this.state;

    return (<ul className="markerList">
      {
        this.props.records.map((marker, id) => (<li key={id} className={`markerBox ${ (selectedId === id)
            ? 'selectedMarker'
            : ''}`} onClick={() => this.onSelect(id)}>
          <div className="markerBox__name">
            <span>{marker.name}</span>
          </div>
          <div className="markerBox__icon">
            <img src={`http://localhost:8080/images/${marker.icon}`} alt={marker.icon} />
          </div>
        </li>))
      }
    </ul>);
  }
}
const mapStateToProps = (state) => ({
  records: state.marker.records
})

const mapDispatchToProps = {
  fetchMarkerRecord
}

export default connect(mapStateToProps, mapDispatchToProps)(MarkerList);
