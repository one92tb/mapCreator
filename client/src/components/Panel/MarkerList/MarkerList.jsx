import React, { Component } from "react";
import "./markerList.css";
import { fetchRecords } from "../../../actions/fetchRecords";
import { getSelectedMarker } from "../../../actions/getSelectedMarker";
import { connect } from "react-redux";

class MarkerList extends Component {
  constructor(props) {
    super(props);
    this.state = {
      selectedId: ""
    };
  }

  componentDidMount() {
    this.props.fetchRecords();
  }

  onSelect = (marker, id) => {
    const getSelectedMarker = this.props.getSelectedMarker;

    if (marker.id === this.state.selectedId) {
      id = "";
      this.setState({ selectedId: id });
      getSelectedMarker({
        id: undefined,
        name: "",
        url: "IMG-default.png"
      });
    } else {
      this.setState({ selectedId: id });
      getSelectedMarker({
        id: marker.id,
        name: marker.name,
        icon: marker.icon,
        url: `http://localhost:8080/images/${marker.icon}`
      });
    }
  };

  render() {
    const selectedId = this.state.selectedId;

    return (
      <ul className="markerList">
        {this.props.records.map((marker, id) => (
          <li
            key={marker.id}
            className={`markerBox ${
              selectedId === marker.id // klikniete id z recordem z map
                ? "selectedMarker"
                : ""
            }`}
            onClick={() => this.onSelect(marker, marker.id)}
          >
            <div className="markerBox__name">
              <span>{marker.name}</span>
            </div>
            <div className="markerBox__icon">
              <img
                src={`http://localhost:8080/images/${marker.icon}`}
                alt={marker.icon}
              />
            </div>
          </li>
        ))}
      </ul>
    );
  }
}

const mapStateToProps = state => ({
  records: state.marker.records,
  isRemoved: state.marker.isRemoved
});

const mapDispatchToProps = {
  fetchRecords,
  getSelectedMarker
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(MarkerList);
