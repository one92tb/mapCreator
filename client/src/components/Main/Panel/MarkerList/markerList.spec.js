import React from "react";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";
import configureStore from "redux-mock-store";
import { MarkerList } from "./MarkerList";

describe("marker list component", () => {
  let wrapper;

  it("passess all props to markerList component", () => {
    const fetchMarkers = jest.fn();
    const getSelectedMarker = jest.fn();
    const disableMarkers = jest.fn();
    const markers = [
      {
        icon: "1548784752159.png",
        id: 1,
        name: "test1",
        userId: 1
      },
      {
        icon: "1550173679016.png",
        id: 2,
        name: "basen",
        userId: 1
      }
    ];
    const selectedMarker = {
      icon: "1548784752159.png",
      id: 1,
      name: "test1",
      userId: 1,
      url: "http://localhost:8080/images/1548784752159.png"
    };

    const isNavSelect = true;

    wrapper = shallow(
      <MarkerList
        markers={markers}
        selectedMarker={selectedMarker}
        fetchMarkers={fetchMarkers}
        getSelectedMarker={getSelectedMarker}
        disableMarkers={disableMarkers}
        isNavSelect={isNavSelect}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it("componentDidMount", () => {
    const fetchMarkers = jest.fn();
    const getSelectedMarker = jest.fn();
    const disableMarkers = jest.fn();
    const markers = [
      {
        icon: "1548784752159.png",
        id: 1,
        name: "test1",
        userId: 1
      },
      {
        icon: "1550173679016.png",
        id: 2,
        name: "basen",
        userId: 1
      }
    ];
    const selectedMarker = {
      icon: "1548784752159.png",
      id: 1,
      name: "test1",
      userId: 1,
      url: "http://localhost:8080/images/1548784752159.png"
    };

    wrapper = shallow(
      <MarkerList
        markers={markers}
        selectedMarker={selectedMarker}
        fetchMarkers={fetchMarkers}
        getSelectedMarker={getSelectedMarker}
        disableMarkers={disableMarkers}
        isNavSelect={true}
      />
    );

    expect(fetchMarkers).toHaveBeenCalledTimes(1);
  });

  it("componentDidUpdate", () => {
    const fetchMarkers = jest.fn();
    const getSelectedMarker = jest.fn();
    const disableMarkers = jest.fn();
    const markers = [
      {
        icon: "1548784752159.png",
        id: 1,
        name: "test1",
        userId: 1
      },
      {
        icon: "1550173679016.png",
        id: 2,
        name: "basen",
        userId: 1
      }
    ];

    wrapper = shallow(
      <MarkerList
        markers={markers}
        selectedMarker={""}
        fetchMarkers={fetchMarkers}
        getSelectedMarker={getSelectedMarker}
        disableMarkers={disableMarkers}
        isNavSelect={true}
      />
    );

    wrapper.setState({
      selectedId: 5
    });

    expect(wrapper.instance().state.selectedId).toBe("");
  });

  it("onSelect - select marker", () => {
    const fetchMarkers = jest.fn();
    const getSelectedMarker = jest.fn(selectedMarker => {
      wrapper.setProps({
        selectedMarker: selectedMarker
      });
    });
    const disableMarkers = jest.fn();
    const markers = [
      {
        icon: "1548784752159.png",
        id: 1,
        name: "test1",
        userId: 1
      },
      {
        icon: "1550173679016.png",
        id: 2,
        name: "basen",
        userId: 1
      }
    ];

    const expectedSelectedMarker = {
      icon: "1548784752159.png",
      id: 1,
      name: "test1",
      userId: 1,
      url: "http://localhost:8080/images/1548784752159.png"
    };
    const isNavSelect = true;

    wrapper = mount(
      <MarkerList
        markers={markers}
        selectedMarker={""}
        fetchMarkers={fetchMarkers}
        getSelectedMarker={getSelectedMarker}
        disableMarkers={disableMarkers}
        isNavSelect={isNavSelect}
      />
    );

    const li = wrapper.find("li").first();
    li.simulate("click");

    expect(getSelectedMarker).toHaveBeenCalledTimes(1);
    expect(wrapper.instance().state.selectedId).toBe(1);
    expect(wrapper.instance().props.selectedMarker).toEqual(
      expectedSelectedMarker
    );
    expect(isNavSelect).toBe(true);
  });

  it("onSelect - unselect marker", () => {
    const fetchMarkers = jest.fn();
    const getSelectedMarker = jest.fn(selectedMarker => {
      wrapper.setProps({
        selectedMarker: selectedMarker
      });
    });
    const disableMarkers = jest.fn();
    const markers = [
      {
        icon: "1548784752159.png",
        id: 1,
        name: "test1",
        userId: 1
      },
      {
        icon: "1550173679016.png",
        id: 2,
        name: "basen",
        userId: 1
      }
    ];

    const selectedMarker = {
      icon: "1548784752159.png",
      id: 1,
      name: "test1",
      userId: 1,
      url: "http://localhost:8080/images/1548784752159.png"
    };
    const isNavSelect = true;

    wrapper = mount(
      <MarkerList
        markers={markers}
        selectedMarker={selectedMarker}
        fetchMarkers={fetchMarkers}
        getSelectedMarker={getSelectedMarker}
        disableMarkers={disableMarkers}
        isNavSelect={isNavSelect}
      />
    );

    wrapper.setState({
      selectedId: 1
    });

    const expectedSelectedMarker = {
      id: undefined,
      name: "",
      url: "IMG-default.png"
    };

    const li = wrapper.find("li").first();
    li.simulate("click");

    expect(getSelectedMarker).toHaveBeenCalledTimes(1);
    expect(wrapper.instance().state.selectedId).toBe("");
    expect(wrapper.instance().props.selectedMarker).toEqual(
      expectedSelectedMarker
    );
  });

  it("onSelect - add marker to filteredMarkers in state", () => {
    const fetchMarkers = jest.fn();
    const getSelectedMarker = jest.fn(selectedMarker => {
      wrapper.setProps({
        selectedMarker: selectedMarker
      });
    });
    const disableMarkers = jest.fn();
    const markers = [
      {
        icon: "1548784752159.png",
        id: 1,
        name: "test1",
        userId: 1
      },
      {
        icon: "1550173679016.png",
        id: 2,
        name: "basen",
        userId: 1
      }
    ];

    const isNavSelect = false;

    wrapper = mount(
      <MarkerList
        markers={markers}
        selectedMarker={""}
        fetchMarkers={fetchMarkers}
        getSelectedMarker={getSelectedMarker}
        disableMarkers={disableMarkers}
        isNavSelect={isNavSelect}
      />
    );

    const li = wrapper.find("li").first();
    li.simulate("click");

    expect(disableMarkers).toHaveBeenCalledTimes(1);
    expect(wrapper.state().filteredMarkers).toEqual([markers[0]]);
  });
  it("onSelect - remove marker from filteredMarkers in state", () => {
    const fetchMarkers = jest.fn();
    const getSelectedMarker = jest.fn(selectedMarker => {
      wrapper.setProps({
        selectedMarker: selectedMarker
      });
    });
    const disableMarkers = jest.fn();
    const markers = [
      {
        icon: "1548784752159.png",
        id: 1,
        name: "test1",
        userId: 1
      },
      {
        icon: "1550173679016.png",
        id: 2,
        name: "basen",
        userId: 1
      }
    ];

    const isNavSelect = false;

    wrapper = mount(
      <MarkerList
        markers={markers}
        selectedMarker={""}
        fetchMarkers={fetchMarkers}
        getSelectedMarker={getSelectedMarker}
        disableMarkers={disableMarkers}
        isNavSelect={isNavSelect}
      />
    );

    wrapper.setState({
      filteredMarkers: [markers[0]]
    });

    const li = wrapper.find("li").first();
    li.simulate("click");

    expect(disableMarkers).toHaveBeenCalledTimes(1);
    expect(wrapper.state().filteredMarkers).toEqual([]);
  });
});
