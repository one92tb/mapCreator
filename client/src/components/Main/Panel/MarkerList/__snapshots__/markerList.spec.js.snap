// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`marker list component passess all props to markerList component 1`] = `
<div>
  <li
    isSelected={false}
    key="1"
    onClick={[Function]}
  >
    <div>
      <img
        alt="1548784752159.png"
        src="http://localhost:8080/images/1548784752159.png"
      />
    </div>
    <span>
      test1
    </span>
  </li>
  <li
    isSelected={false}
    key="2"
    onClick={[Function]}
  >
    <div>
      <img
        alt="1550173679016.png"
        src="http://localhost:8080/images/1550173679016.png"
      />
    </div>
    <span>
      basen
    </span>
  </li>
</div>
`;
